edk2toollib/uefi/telem/parser_main.py

```diff
@@ -0,0 +1,660 @@
+Revision            (8  byte offset, 2  byte length) : Represents the major and minor version number of the Error Record definition
+Validation Bits     (10 byte offset, 1  byte length) : Indicates the validity of the FRU ID and FRU String fields
+Reserved            (11 byte offset, 1  byte length) : Must be zero
+Flags               (12 byte offset, 4  byte length) : Contains info describing the Error Record (ex. Is this the primary section of the error, has the component been reset if applicable, has the error been contained)
+Section Type        (16 byte offset, 16 byte length) : Holds a pre-defined GUID value indicating that this section is from a particular error
+FRU ID              (32 byte offset, 16 byte length) : ? Not detailed in the CPER doc
+Section Severity    (48 byte offset, 4  byte length) : Value from 0 - 3 indicating the severity of the error
+FRU String          (52 byte offset, 20 byte length) : String identifying the FRU hardware
+
+
+Python Struct Format Characters
+
+Format      C Type                  Python Type     Standard Size
+x           pad byte                none            1
+c           char                    integer         1
+b           signed char             integer         1
+B           unsigned char           integer         1
+?           _Bool                   bool            1
+h           short                   integer         2
+H           unsigned short          integer         2
+i           int                     integer         4
+I           unsigned int            integer         4
+l           long                    integer         4
+L           unsigned long           integer         4
+q           long long               integer         8
+Q           unsigned long long      integer         8
+n           ssize_t                 integer
+N           size_t                  integer
+e           (6)                     float           2
+f           float                   float           4
+d           double                  float           8
+s           char[]                  bytes
+p           char[]                  bytes
+P           void *                  integer
+"""
+
+CPER_HEADER_SIZE            = 128 # A CPER header is 128 bytes 
+CPER_SECTION_HEADER_SIZE    = 72  # A CPER section header is 72 bytes
+FriendlyNames               = {}  # dict loaded from friendlynames.csv used to associate guids with friendly strings
+Parsers                     = []  # list of plugin classes which inherit from CPER_SECTION_DATA and are therefore capable of parsing section data
+
+# Using these lambda concat statements to guarantee linear runtime when concatonating strings, and for simplicity
+concat_s = lambda x,y  : (''.join([x,y," "]))  # Concatenate two strings and put a space at the end
+concat_n = lambda x,y  : (''.join([x,y,"\n"])) # Concatenate two strings and put a newline at the end
+concat   = lambda x,y  : (''.join([x,y]))      # Concatenate two strings
+
+# Useful lambdas
```

Comment:
I think you could make a case for these functions to be inside utility functions since they look quite useful